api/src/poem_backend/log.rs

```diff
@@ -0,0 +1,75 @@
+// Copyright (c) Aptos
+// SPDX-License-Identifier: Apache-2.0
+
+use std::time::Duration;
+
+use crate::metrics::RESPONSE_STATUS;
+use aptos_logger::{
+    debug, error,
+    prelude::{sample, SampleRate},
+    sample::Sampling,
+    Schema,
+};
+use poem::{http::header, Endpoint, Request, Response, Result};
+
+/// Logs information about the request and response if the response status code
+/// is >= 500, to help us debug since this will be an error on our side.
+/// We also do general logging of the status code alone regardless of what it is.
+pub async fn middleware_log<E: Endpoint>(next: E, request: Request) -> Result<Response> {
+    let start = std::time::Instant::now();
+
+    let mut log = HttpRequestLog {
+        remote_addr: request.remote_addr().as_socket_addr().cloned(),
+        method: request.method().to_string(),
+        path: request.uri().path().to_string(),
+        status: 0,
+        referer: request
+            .headers()
+            .get(header::REFERER)
+            .and_then(|v| v.to_str().ok().map(|v| v.to_string())),
+        user_agent: request
+            .headers()
+            .get(header::USER_AGENT)
+            .and_then(|v| v.to_str().ok().map(|v| v.to_string())),
+        elapsed: Duration::from_secs(0),
+        forwarded: request
+            .headers()
+            .get(header::FORWARDED)
+            .and_then(|v| v.to_str().ok().map(|v| v.to_string())),
+    };
+
+    let response = next.get_response(request).await;
+
+    let elapsed = start.elapsed();
+
+    log.status = response.status().as_u16();
+    log.elapsed = elapsed;
+
+    if log.status >= 500 {
+        sample!(SampleRate::Duration(Duration::from_secs(1)), error!(log));
+    } else {
+        debug!(log);
+    }
+
+    RESPONSE_STATUS
+        .with_label_values(&[log.status.to_string().as_str()])
+        .observe(elapsed.as_secs_f64());
+
+    Ok(response)
+}
+
+// TODO: Figure out how to have certain fields be borrowed, like in the
+// original implementation.
```

Comment:
These should be able to all be borrowed if it's using the one from the Aptos logger.  But it's not high priority.